Extends TextBoxRegex with ValidationMode.Instantly/Number

jp-weber opened the pull request:
Issue: #1757


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build or CI related changes
[ ] Documentation content changes
[ ] Sample app changes
[ ] Other... Please describe:
```


## What is the current behavior?
Currently it is not possible to validate the textbox instantly at input. Especially with numbers the current behavior is not fully taken into account, because an incorrect input is validated only after lost of focus (ValidationMode.Focus) and the TextBox value will be delete. 

With the current behavior the user can use control elements (with focus of the textbox), although the input is not validated.




## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [x] Docs have been added/updated which fit [documentation template](https://raw.githubusercontent.com/Microsoft/UWPCommunityToolkit/tree/master/docs/.template.md). (for bug fixes / features)
- [x] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/Microsoft/UWPCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)


## What is the new behavior?
With the expansion there are two new validation modes, "Instantly" and "InstantlyNumber".

Instantly:  Update IsValid property instantly with validation result at input and in case the textbox is not valid clear its value.

InstantlyNumber: Update IsValid property instantly with validation result and clear the newest character at input which is not valid.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/Microsoft/UWPCommunityToolkit?pullRequest=1776) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/Microsoft/UWPCommunityToolkit?pullRequest=1776) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/Microsoft/UWPCommunityToolkit?pullRequest=1776) it.</sub>

nmetulev commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.Data.cs:
```diff
@@ -34,7 +34,17 @@ public enum ValidationMode
+            /// <summary>
+            /// Update IsValid property instantly with validation result and clear the newest character at input which is not valid
+            /// </summary>
+            InstantlyNumber
```
Not sure if this is the right name for this. How about something like Dynamic or Latest

jp-weber replied:
Okay, I changed the name to Dynamic.

nmetulev commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.Data.cs:
```diff
@@ -34,7 +34,17 @@ public enum ValidationMode
             /// <summary>
             /// Update IsValid property with validation result and in case the textbox is not valid clear its value
```
Update comment to specify the value is cleared on focus lost

jp-weber replied:
Done.

skendrot commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -54,7 +54,15 @@ private static void TextBoxRegexPropertyOnChange(DependencyObject sender, Depend
         {
             var textbox = (TextBox)sender;
-            ValidateTextBox(textbox, false);
+            var validationMode = (ValidationMode)textbox.GetValue(ValidationModeProperty);
```
Remove the if and pass the value of the if statement to the second parameter
`ValidateTextBox(textbox, (validationMode == ValidationMode.Instantly || validationMode == ValidationMode.InstantlyNumber));`

jp-weber replied:
Done.

skendrot commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -109,10 +117,21 @@ private static void ValidateTextBox(TextBox textbox, bool force = true)
                 {
-                    var validationModel = (ValidationMode)textbox.GetValue(ValidationModeProperty);
-                    if (validationModel == ValidationMode.Forced)
+                    if (textbox.Text != string.Empty)
```
use `string.IsNullOrEmpty`

jp-weber replied:
Done.

skendrot commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -109,10 +117,21 @@ private static void ValidateTextBox(TextBox textbox, bool force = true)
+                        {
+                            textbox.Text = string.Empty;
+                        }
+                        else if (validationModel == ValidationMode.InstantlyNumber)
```
Would BeforeTextChanging be beneficial for this? That way you don't have to change the text.

jp-weber replied:
Yes, I will look into BeforeTextChanging. The question is, the event needs the newest SDK? What is the best practise for handle that?
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textboxbeforetextchangingeventargs

skendrot replied:
Could check if the contract is present and act accordingly. I think we could do that in a separate PR though. @nmetulev

skendrot commented:
Looking good. Need to update the docs

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.Data.cs:
```diff
@@ -17,7 +17,9 @@ namespace Microsoft.Toolkit.Uwp.UI.Extensions
     /// If <see cref="ValidationMode"> is set to Normal then IsValid will be set according to whether the regex is valid.</see>
-    /// If <see cref="ValidationMode"> is set to Forced and the input is not valid the TextBox text will be cleared.</see>
+    /// If <see cref="ValidationMode"> is set to Forced, the input will be validated if the TextBox loses focus, if it is invalid, the TextBox text will be deleted.</see>
+    /// If <see cref="ValidationMode"> is set to Instantly, the input will be validated immediately, if it is invalid, the text of the TextBox will be deleted.</see>
```
Is this useful?  If the user types something wrong it'll erase all their text, no?

I think the 'Dynamic' mode is more standard?

Also, have we checked how this behaves with Undo?

jp-weber replied:
@michael-hawker I agree, but I didn't want to change with this PR the extension, only expand. If it is wanted, I would delete force and instantly.

I debuged it and it works with undo and copy&paste.

nmetulev replied:
I agree that instantly seems unnecessary, I would not want to have to retype something just because i miss-typed one character.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -109,10 +110,21 @@ private static void ValidateTextBox(TextBox textbox, bool force = true)
+                        }
+                        else if (validationModel == ValidationMode.Dynamic)
+                        {
+                            textbox.Text = textbox.Text.Remove(textbox.Text.Length - 1);
```
Can't they insert a character in the middle of the text?

nmetulev commented:
ping @jp-weber

jp-weber commented:
@nmetulev yes, I will insert the necessary documentation this weekend.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -102,22 +104,18 @@ private static void ValidateTextBox(TextBox textbox, bool force = true)
-                    if (!string.IsNullOrEmpty(textbox.Text))
+                    textbox.Text = string.Empty;
+                }
+                if (!string.IsNullOrEmpty(textbox.Text))
```
else if

jp-weber replied:
Done.

michael-hawker commented:
The Sample App page should be updated to add a couple of new boxes to show the use cases for these new properties.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -102,22 +104,18 @@ private static void ValidateTextBox(TextBox textbox, bool force = true)
-                            textbox.Text = string.Empty;
-                        }
-                        else if (validationModel == ValidationMode.Dynamic)
+                        textbox.Text = textbox.Text.Remove(textbox.Text.Length - 1);
```
You should use the textbox selection info in order to erase the character at its insertion point, it's a bit odd for the rest of the data to be axed if the user inserts something not at the end.

jp-weber replied:
Done. But if I insert a string which is invalid, for example "34234f34234", the text will be deleted up to "34234".

michael-hawker commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.cs:
```diff
@@ -23,7 +23,9 @@ namespace Microsoft.Toolkit.Uwp.UI.Extensions
     /// If <see cref="ValidationMode"> is set to Normal then IsValid will be set according to whether the regex is valid.</see>
-    /// If <see cref="ValidationMode"> is set to Forced and the input is not valid the TextBox text will be cleared.</see>
+    /// If <see cref="ValidationMode"> is set to Forced, the input will be validated if the TextBox loses focus, if it is invalid, the TextBox text will be deleted.</see>
+    /// If <see cref="ValidationMode"> is set to Instantly, the input will be validated immediately, if it is invalid, the text of the TextBox will be deleted.</see>
```
This still seems like a very specific and user unfriendly behavior.  With the Dynamic mode and IsValid a developer could implement this behavior if they really wanted it easily.

nmetulev commented:
ping @jp-weber, do you think you can address the requested changes by tomorrow (code freeze for 2.2)

jp-weber commented:
@nmetulev sorry, in the week I have not the time.

nmetulev commented:
No worries, moved it to 3.0 :)

jp-weber commented:
@nmetulev I changed the ValidationMode and deleted force and instantly. I implement a use case for the problem with Email/PhoneNumber and the Dynamc ValidationMode. Because an input is not possible when the Email/PhoneNumber is not valid and the newest character will be deleted.

One problem is still open, when the ValidationMode is Dynamic and the ValidationType is Custom with a complex Regex like Email.

nmetulev commented on Microsoft.Toolkit.Uwp.UI/Extensions/TextBoxRegEx/TextBoxRegex.Properties.cs:
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.Toolkit.Uwp.UI.Extensions
     /// <remarks>
     /// If <see cref="ValidationMode"> is set to Normal then IsValid will be set according to whether the regex is valid.</see>
-    /// If <see cref="ValidationMode"> is set to Forced and the input is not valid the TextBox text will be cleared.</see>
+    /// If <see cref="ValidationMode"> is set to Dynamic, the input will be validated at text changed. If the newest charachter is invalid, only invalid character of the Textbox will be deleted.</see>
```
Typo with character. Same in other files (not sure if you need this in every single file btw)

nmetulev commented:
ping @jp-weber

nmetulev commented:
@jp-weber, this looks good. I think the only thing missing is the documentation

nmetulev commented:
@jp-weber, we have code freeze this Wednesday for next release, are you able to update the docs before then so we can include this?

jp-weber commented:
@nmetulev sure, I will commit it today.

nmetulev commented on docs/extensions/TextBoxRegex.md:
```diff
@@ -11,30 +11,24 @@ The [TextBoxRegex Property](https://docs.microsoft.com/dotnet/api/microsoft.tool
-2) Dynamic : This type extends 1) Normal and if is the newest input of the Textbox  is invalid, the character which is invalied will be deleted.
-
-
-Main Attached Properties:
```
Why was this deleted?

jp-weber replied:
Thought this was updated in the new version here? 
https://github.com/Microsoft/WindowsCommunityToolkit/blob/master/docs/extensions/TextBoxRegex.md#attached-properties
So I have deleted the old code.
